### PR TITLE
Edit Entity Mapping View

### PR DESCRIPTION
**Pull request by tommartensen**
Yet to be tested with real backend and refactored (code duplication).

### CODE REVIEW

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 1]
```diff
@@ -39,6 +45,22 @@ class CreateEntityMappingView extends ConnectionComponent {
 		}];
 	}
 
+	static getLoadedStateFromResponse(entityMapping) {
+		let mappings = [];
```
Comment (hpibptlab) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: MINOR Make "mappings" "const". [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS3353)

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 2]
```diff
@@ -249,15 +304,140 @@ class CreateEntityMappingView extends ConnectionComponent {
...
+	}
+
 	render() {
+		if (!this.isCreateView) {
+			return this.renderEditView();
+		}
+		return this.renderCreateView();
+	}
+
+	renderEditView() {
```
Comment (hpibptlab) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: MAJOR Function has a complexity of 11 which is greater than 10 authorized. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AFunctionComplexity)

**Comment by hpibptlab** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 3]
```diff
@@ -23,64 +24,121 @@ class CreateEntityMappingView extends ConnectionComponent {
...
-			this.props.createEntityMapping({
-				EventTypeId: this.state.selectedEventTypeId.value,
-				EntityTypeId: this.state.selectedEntityTypeId.value,
-				TargetStatus: this.state.targetStatus,
-				EventEntityMappingConditions: entityMappingConditions
+			isValid = false;
+		}
+		if (this.state.selectedEventTypeId.value === null) {
+			this.setState({
+				selectedEventTypeId: {value: null, errorMessage: config.messages.requiredFieldMessage}
```
Comment (MaximilianV) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: `selectedEventTypeId` seems to be an object and not an int? Maybe the naming is not appropriate.

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 4]
```diff
@@ -249,98 +311,168 @@ class CreateEntityMappingView extends ConnectionComponent {
 			</div>);
 	}
 
-	render() {
+	renderComponentBody(initialData, optionalActions, submitCallback) {
```
Comment (MaximilianV) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: Can we split this _very_ long further?

**Review by MaximilianV — CHANGES_REQUESTED** [review]

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 3 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: The required change was addressed.

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 5]
```diff
@@ -13,74 +13,129 @@ import ErrorMessage from './../Utils/ErrorMessage.js';
...
 		super(props);
 		this.state = {
 			targetStatus: "",
-			selectedEventTypeId: {value: null, errorMessage: ""},
-			selectedEntityTypeId: {value: null, errorMessage: ""},
+			selectedEventType: {value: null, errorMessage: ""},
+			selectedEntityType: {value: null, errorMessage: ""},
 			mappings: CreateEntityMappingView.getDefaultMappings()
 		};
+		this.isCreateView = typeof this.props.match.params.entityMappingId === 'undefined';
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: This is not very nice.
Why not using props?

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 6]
```diff
@@ -13,74 +13,129 @@ import ErrorMessage from './../Utils/ErrorMessage.js';
...
-			const  entityMappingConditions = this.state.mappings.map((mappingStatement) => {
-				return ({
-					EventTypeAttributeId: mappingStatement.eventTypeAttribute.value,
-					EntityTypeAttributeId: mappingStatement.entityTypeAttribute.value
-				});
+	isValidInput() {
+		let isValid = true;
+		if(this.state.selectedEntityType.value === null) {
+			this.setState({
+				selectedEntityTypeId: {value: null, errorMessage: config.messages.requiredFieldMessage}
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: You removed the id from the name before. Are you sure that "id" is the suitable solution here?

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 7]
```diff
@@ -139,70 +183,53 @@ class CreateEntityMappingView extends ConnectionComponent {
...
-			errorText={this.state.mappings[key].entityTypeAttribute.errorMessage}
-			onChange={(event, index, value) => {
-                this.handleEntityTypeAttributeChange(key, value);
-            }}>
-            {CreateEntityMappingView.getMenuItems(entityTypeAttributes)}
-		</SelectField>);
-    }
+		return (
+			<SelectField
+				floatingLabelText="Select Event Type Attribute"
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: Please export text to config

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 8]
```diff
@@ -139,70 +183,53 @@ class CreateEntityMappingView extends ConnectionComponent {
...
+	getEntityTypeAttributesDropDown(key, selectedEntityTypeAttribute) {
+		const entityTypeAttributes = this.props.entityTypeAttributes.value;
+		return (
+			<SelectField
+				floatingLabelText="Select Entity Type Attribute"
+				fullWidth={true}
+				value={selectedEntityTypeAttribute}
+				errorText={this.state.mappings[key].entityTypeAttribute.errorMessage}
+				onChange={(event, index, value) => {
+					this.handleEntityTypeAttributeChange(key, value);
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js:217]: Binding would be more beautiful

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 9]
```diff
@@ -215,7 +242,43 @@ class CreateEntityMappingView extends ConnectionComponent {
 		this.setState({mappings});
 	}
 
-	loadAttributes() {
+	// is called to load the existing entity mapping in the edit view
+	entityMappingPending() {
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: is never called

Comment (julianweise) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: naming is not very good => why pending? isn´t it used to load data into the view?
A name like loadDataIntoEditView or something like that would be sufficient.

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 10]
```diff
@@ -249,98 +312,192 @@ class CreateEntityMappingView extends ConnectionComponent {
...
+						floatingLabelText="Select Entity Type"
+						fullWidth={true}>
+						{CreateEntityMappingView.getMenuItems(initialData.entityTypes)}
+					</SelectField>
+				</Col>
+			</Row>
+		);
+	}
+
+	static renderAttributeFields(initialData) {
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: Implementing a static method for only accessing a member variable is useless

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 11]
```diff
@@ -249,98 +312,192 @@ class CreateEntityMappingView extends ConnectionComponent {
...
-		if(optionalActions && optionalActions.fulfilled) {
-			window.history.back();
-			return null;
-		}
-		const entityTypes = this.transformHierarchy(this.props.entityTypeHierarchy.value);
+		const entityTypes = CreateEntityMappingView.transformHierarchy(this.props.entityTypeHierarchy.value);
 		const connectionIncomplete = super.render(allFetches);
 		if (connectionIncomplete) {
-			return connectionIncomplete;
+			return false;
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: I am not sure you should return false here. 
connectionIncomplete calls the render method of ConnectionComponent (via super) and returns either a Loading animation or a ErrorMessage if the loading procedure has not finished yet.

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 12]
```diff
@@ -249,98 +312,192 @@ class CreateEntityMappingView extends ConnectionComponent {
...
+
+	renderEditView() {
+		const optionalActions = this.props.updateEntityMappingResponse;
+		CreateEntityMappingView.handleOptionalActionsSuccess(optionalActions);
+		const initialData = this.fetchInitialData();
+		if (!initialData) {
+			return <LoadingAnimation/>;
+		}
+		return (
+			<div ref={() => {this.oldValuesShouldBeLoaded = true;}}>
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: You should use the PromiseState responses (e.g.: fulfilled/rejected/pending) to decided whether fetched-data-dependent-UI has been rendered.

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 13]
```diff
@@ -249,98 +312,192 @@ class CreateEntityMappingView extends ConnectionComponent {
...
+				<Header title={"Edit Entity Mapping"}/>
+				<div className={AppStyles.elementMarginTop}>
+					{this.renderComponentBody(initialData, optionalActions, this.submitUpdatedMapping)}
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: ComponentBody: Insufficient naming
Component refers to entities of react framework ...

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 14]
```diff
@@ -249,98 +312,192 @@ class CreateEntityMappingView extends ConnectionComponent {
...
+	componentWillMount() {
+		if (!this.isCreateView) {
+			this.props.lazyLoadEntityMapping();
+		}
+	}
+
+	componentDidUpdate(prevProps, prevState) {
+		if (!this.isCreateView && !this.oldValuesWereLoaded && this.oldValuesShouldBeLoaded) {
+			// in edit view, when old values should be loaded into the interface
+			const entityMapping = this.props.entityMapping.value;
```
Comment (julianweise) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: Why don´t you check for PromiseState?

**Review by julianweise — DISMISSED** [review]

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 5 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: We cannot use props, as the view is accessed via url from the settings view and cannot be passed properties. We also can't pass this from the settings view, as we cannot read the entityMappingId from the URL when rendering a component in a route.

Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: However, I will add an explaining comment.

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 6 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: True that, forgot the remove it here.

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 9 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: was removed

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 13 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: --> renderMappingForm

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 7 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: if @frederike-ramin does it as well....

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 10 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: done

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 8 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js:217]: As discussed: binding is equally bad

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 14 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: yes, that might be better

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 11 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: the render of the fetching method is returned

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 12 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: done

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 15]
```diff
@@ -249,98 +296,189 @@ class CreateEntityMappingView extends ConnectionComponent {
...
+
+	// loads entity mapping in edit view
+	componentWillMount() {
+		if (!this.isCreateView) {
+			this.props.lazyLoadEntityMapping();
+		}
+	}
+
+	componentDidUpdate(prevProps, prevState) {
+		if (!this.isCreateView && !this.oldValuesWereLoaded && this.oldValuesShouldBeLoaded) {
```
Comment (hpibptlab) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: MAJOR Merge this if statement with the nested one. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3ACollapsibleIfStatements)

File: src/CreateEntityMappingView/CreateEntityMappingView.js [hunk 16]
```diff
@@ -19,68 +18,123 @@ class CreateEntityMappingView extends ConnectionComponent {
...
-	static getMenuItems(items) {
-		return items.map(
-			(item, key) => {
-				return <MenuItem
-					key={key}
-					value={item.Id}
-					primaryText={item.Name}/>;
+	// returns the state loaded from a response
+	static getLoadedStateFromResponse(entityMapping) {
+		let mappings = [];
```
Comment (hpibptlab) [inline src/CreateEntityMappingView/CreateEntityMappingView.js]: MINOR Make "mappings" "const". [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS3353)

File: src/CreateEntityMappingView/CreateEntityMappingView.js (hunk 4 above)
Comment (tommartensen) [reply src/CreateEntityMappingView/CreateEntityMappingView.js]: This _very_ long method was split.

**Review by MaximilianV — APPROVED** [review]

**Pull request merged**